llvm/lib/Target/X86/X86ZakharovInstCounterPass.cpp

```diff
@@ -0,0 +1,39 @@
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+namespace {
+
+class X86ZakharovInstCounterPass : public MachineFunctionPass {
+public:
+  static char ID;
+  X86ZakharovInstCounterPass() : MachineFunctionPass(ID) {}
+  bool runOnMachineFunction(llvm::MachineFunction &MF) override {
+    auto *GV = MF.getFunction().getParent()->getNamedGlobal("ic");
+    if (!GV) {
+      return false;
+    }
+    auto *TIF = MF.getSubtarget().getInstrInfo();
+    MachineInstrBuilder MIB = BuildMI(MF.back(), MF.back().back(), DebugLoc(),
+                                      TIF->get(X86::MOV64mi32));
+    MIB.addReg(0);
+    MIB.addImm(1);
+    MIB.addReg(0);
+    MIB.addGlobalAddress(GV);
+    MIB.addReg(0);
+    MIB.addImm(MF.getInstructionCount());
```

Comment:
Please, notice that this is number of instructions in MiR program representation. Whilst, number of instructions, executed during program run, are required.

Reply:
Ok. Let me explain in another words. If your block corresponds to a loop and has 5 instructions in it your pass increments counter on 4 instructions (by the way, terminator is instruction too and it does a work as well). Whilst, a loop may have 100 iterations during execution. So, 100 * 5 instructions will be performed during runtime.

Reply:
But now I'm incrementing the counter at the end of each base block, and when the loop executes, the counter will increment when each base block in the loop is executed.
```
for (auto &MBB : MF) {
      int Counter = std::distance(MBB.begin(), MBB.end());
      auto Place = MBB.getFirstTerminator();
      if (Place != MBB.begin()) {
        --Place;
      }
      BuildMI(MBB, Place, DL, TII->get(X86::ADD64mi32))
          .addReg(0)
          .addImm(1)
          .addReg(0)
          .addGlobalAddress(GV)
          .addReg(0)
          .addImm(Counter);
    }
```

Reply:
Yes, sorry, my bad. I read it inadvertently.